students/rusty_mann/session05/mailroom_session05.py

```diff
@@ -0,0 +1,202 @@
+    return name
+
+
+def get_donor_name():
+    """handle exceptions and return donor name"""
+    while True:
+        name = donor_selection()
+        if name == "list".strip().lower():
+            show_donor_list()
+        elif name == "menu".strip().lower():
+            return None
+        else:
+            try:
+                first, last = split_name(name)
+                name = last + ", " + first
+            except IndexError:
+                print("Please enter a full name: ""Last name"", ""First name"": ")
+            else:
+                return name
+
+
+def donation_selection():
+    amount_str = str(input("Please enter a donation amount (or 'menu' to exit)> "))
+    return amount_str
+
+
+def get_donation_amount():
+    """handle exceptions and return donation amount"""
+    while True:
+        donation = donation_selection()
+        if donation == "menu".strip().lower():
+            return None
+        else:
+            try:
+                amount = float(donation)
+            except ValueError:
+                print("Error: Please enter a number")
+            else:
+                return amount
+
+
+def send_donor_email():
+    """print thank you message to terminal"""
+    name = get_donor_name()
+    if name == None:
+        return None
+    amount = get_donation_amount()
+    if amount == None:
+        return None
+    add_donor(name, amount)
+    donor_dic = make_donor_dict(name, amount)
+    print(make_donor_email(donor_dic))
+
+
+def sort_key(item):
+    """return second item in sequence for sorting donor report"""
+    return item[1]
+
+
+def make_report():
+    """print full donor report to terminal"""
+    rows = []
+    for donor in donor_data:
+        total = sum(donor_data[donor])
+        num = len(donor_data[donor])
+        avg = total / num
+        rows.append((donor, total, num, avg))
+    rows.sort(key=sort_key, reverse=True)
+    print("{:20s}{:15s}{:15s}{:12s}".format(
+        "Donor Name", "|  Total Given", "|  Num Gifts", "|  Average Gift"))
+    print("_" * 67)
+    [print('{:20s}{:15.2f}{:^15d}{:12.2f}'.format(*row)) for row in rows]
+
+
+def quit_program():
+    print("Thank you, come again!")
+    sys.exit(0)
+
+
+def menu_selection():
+    selection = input('''\n
+
+                Would you like to:
+
+                '1' - Send a Thank You 
+                '2' - Create a Report
+                '3' - Send letters to everyone 
+                '4' - Quit
+                > ''')
+    return selection
+
+
+if __name__ == "__main__":
+
+    running = True
+
+    dispatch_dict = {"1": send_donor_email,
+                           "2": make_report,
+                           "3": make_letter_files,
+                           "4": quit_program
+                           }
+
+    while running:
+        response = menu_selection()
+        try:
+            dispatch_dict[response]()
+        except KeyError:
+            print("Error: Please select a valid menu option")
+
```

Comment:
Very well done. Pretty cool that you can stick functions into dictionaries, isn't it?